linkerd/router/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+use ahash::AHashMap;
+use futures::{future, prelude::*};
+use linkerd_error::Error;
+use linkerd_stack::{layer, NewService, NewSpawnWatch, Oneshot, Param, Service, ServiceExt};
+use std::{
+    fmt::Debug,
+    hash::Hash,
+    marker::PhantomData,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tracing::{debug, error};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RouteKeys<K>(Arc<[K]>)
+where
+    K: Eq + Hash + Clone;
+
+pub trait SelectRoute<Req> {
+    type Key: Eq + Hash + Clone + Debug + Send + Sync + 'static;
+    type Error: std::error::Error + Send + Sync + 'static;
```

Comment:
making this trait responsible for returning the error seems like a good call :+1: